### DISCUSSION TITLE
Vitocal 250-SH + Vitodens 300W

### DISCUSSION THREAD

**Discussion by cvbasta75** [open3e/open3e#123] [Q&A]
Guten Morgen,

hätte ein paar Änderungen für die Open3Edatapoints.py.

Datenpunkt 2985: `        2985 : O3EInt16(2, "ExternalHeaterTemperatureSetpoint")`

Datenpunkt 1097+1098: `1097 : O3EComplexType(20, "ElectricityPrice", [O3EInt32(4, "Unknown1"), O3EInt32(4, "Unknown2"), O3EInt32(4, "NormalRate"), O3EInt32(4, "LowRate"), O3EInt32(4, "Unknown3")]), 1098 : O3EComplexType(20, "GasProperties", [O3EInt32(4, "Rate"), O3EInt32(4, "Unknown1"), O3EInt32(4, "Unknown2"), O3EInt32(4, "Unknown3"), O3EInt32(4, "Unknown4")]),`

Die Tarife die da rauskommen müssen allerdings durch 10000 geteilt werden!
Schön wäre es jetzt, wenn man diese auch noch schreiben könnte, dann könnte man automatisch per Skript den aktuellen Tibber-Preis jede Stunde hinterlegen.

**Comment by bu-na** [comment]
Hallo, 

sieht gut aus, würde bei den Energiepreisen die Umrechnung in Cent auch gleich hier erledigen. 
Von der Konvertierung der Unknown-Werte nach Int32 bin ich mit Blick auf die Ergebnisse nicht überzeugt. 
Vielleicht sollte man hier erst mal die Raw-Werte übernehmen, bis uns etwas Besseres einfällt.

Hier mal meine Interpretation zur Diskussion:

    1097 : O3EComplexType(20, "ElectricityPrice", [RawCodec(4, "Unknown1"), RawCodec(4, "Unknown2"), O3EInt32(4, "NormalRate", scale=100), O3EInt32(4, "LowRate", scale = 100), RawCodec(4, "Unknown3")]), # Unit ct
    1098 : O3EComplexType(20, "GasProperties", [O3EInt32(4, "Rate", scale = 100), RawCodec(4, "Unknown1"), RawCodec(4, "Unknown2"), RawCodec(4, "Unknown3"), RawCodec(4, "Unknown4")]), # Unit ct

oder kompakter:
        1098 : O3EComplexType(20, "GasProperties", [O3EInt32(4, "Rate", scale = 100), RawCodec(16, "Unknown")]), # Unit ct

Gruß Bu-Na

**Comment by cvbasta75** [comment]
Bin schon etwas weiter. Hier mal ein PHP Script um die Tibber Preise automatisch in die Vitocal zu bekommen.
Einfach per Cronjob einmal die Stunde ausführen lassen... und schon sind die Preise drinnen.

`<?php

    error_reporting(E_ERROR | E_WARNING);
    setlocale(LC_TIME, "de_DE.utf8");

    $apikey = ""; // Hier kommt euer Tibber API key rein.
    $open3e = "/root/pp/bin/open3e"; // Hier kommt der absolute Pfad zu eurer open3e rein.
    $can = "can0"; // Hier kommt der CAN Bus rein.
    $did = "1097"; // Hier kommt die DID rein, falls abweichend.

    $TimeStamp = date("d.m.Y H:i:s") . " => ";

    $hour = date("H");
    $json = '{"query":"{viewer{homes{currentSubscription{priceInfo{today{total startsAt}}}}}}"}';
    $ch = curl_init('https://api.tibber.com/v1-beta/gql');
    curl_setopt($ch, CURLOPT_URL, 'https://api.tibber.com/v1-beta/gql');
    curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type: application/json', 'Authorization: Bearer '.$apikey));
    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $json);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    $response = curl_exec($ch);
    curl_close($ch);
    $rawdata = json_decode($response, true, 512, JSON_OBJECT_AS_ARRAY);
    $data = $rawdata["data"]["viewer"]["homes"][0]["currentSubscription"]["priceInfo"]["today"];
    if (count($data) == 24) {
        foreach ($data as $rates) {
            $start = substr($rates["startsAt"], 11, 2);
            if ($hour == $start) {
                $price = $rates["total"];
                echo $TimeStamp . "Preis um " . $hour . ": " . $price . "EUR\n";
                $hexprice = dechex(intval(floatval($price) * 10000));
                $hexprice = str_pad($hexprice, 8, "0", STR_PAD_LEFT);
                $part4 = substr($hexprice, 0, 2);
                $part3 = substr($hexprice, 2, 2);
                $part2 = substr($hexprice, 4, 2);
                $part1 = substr($hexprice, 6, 2);
                $newhexprice = $part1 . $part2 . $part3 . $part4;
                echo $TimeStamp . "Preis * 10000 in Hex (lsb...msb): " . $newhexprice . "\n";
                $o3ecom = $open3e . " -c " . $can . " -r " . $did . " --raw";
                $o3eret = exec($o3ecom, $output, $retval);
                if ($retval == 0) {
                    if (strlen($o3eret) == 40) {
                        $unknown1 = substr($o3eret, 0, 16);
                        echo $TimeStamp . "alter RAW-Wert: " . $unknown1;
                        $normalrate = substr($o3eret, 16, 8);
                        echo $normalrate;
                        $lowrate = substr($o3eret, 24, 8);
                        echo $lowrate;
                        $unknown2 = substr($o3eret, 32, 8);
                        echo $unknown2 . "\n";
                        $newrawdata = $unknown1 . $newhexprice . $newhexprice . $unknown2;
                        echo $TimeStamp . "neuer RAW-Wert: " . $newrawdata . "\n";
                        $o3ecom = $open3e . " -c " . $can . " -w " . $did . "=" . $newrawdata . " --raw";
                        $o3eret = exec($o3ecom, $output, $retval);
                        echo $TimeStamp . "Ausgabe open3e: " . $o3eret . "\n";
                        if (strpos($o3eret, "success: True") !== false) {
                            echo $TimeStamp . "Schreiben der RAW Daten erfolgreich.\n";
                        } else {
                            echo $TimeStamp . "FEHLER beim Schreiben der RAW Daten!\n";
                        }
                    } else {
                        echo $TimeStamp . "FEHLER beim Einlesen der RAW Daten!\n";
                    }
                } else {
                    echo $TimeStamp . "FEHLER beim Einlesen der RAW Daten!\n";
                }
            }
        }
    } else {
         echo $TimeStamp . "FEHLER beim Abrufen der Tibber Preise!\n";
    }
    echo "\n";

?>
`

**Comment by cvbasta75** [comment]
Die Regelstrategie lässt sich aber leider nicht schreiben. Hier ändert sich zwar immer nur die letzte Zahl in den Raw-Daten aber:

`/root/pp/bin/open3e -c can0 -r 2404 --raw -v
None 2404 BivalenceControlMode "029600000002"
closing 0x680 - bye!

/root/pp/bin/open3e -c can0 -w 2404=029600000001 --raw -v
write raw: 1664.2404 = 029600000001
Traceback (most recent call last):
  File "/root/pp/bin/open3e", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/root/pp/lib/python3.12/site-packages/open3e/Open3Eclient.py", line 386, in main
    succ,code = dicEcus[ecu].writeByDid(didkey, didVal, raw=True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/pp/lib/python3.12/site-packages/open3e/Open3Eclass.py", line 162, in writeByDid
    response = self.uds_client.write_data_by_identifier(did, val)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/pp/lib/python3.12/site-packages/udsoncan/client.py", line 167, in decorated
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/pp/lib/python3.12/site-packages/udsoncan/client.py", line 521, in write_data_by_identifier
    response = self.send_request(req)
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/root/pp/lib/python3.12/site-packages/udsoncan/client.py", line 2258, in send_request
    raise NegativeResponseException(response)
udsoncan.exceptions.NegativeResponseException: WriteDataByIdentifier service execution returned a negative response ConditionsNotCorrect (0x22)`

**Comment by MyHomeMyData** [comment]
@cvbasta75 @bu-na Ich habe 1097 und 1098 mit der Einheit ct in den Branch develop übernommen, die "unknown" als Raw. Bei 2985 habe ich noch ein `sigend=True` ergänzt, da das bei Temperatursollwerten meistens sinnvoll ist.

**Reply by bu-na** [reply]
Völlig Richtig, da Setpoints idR in 1/10-Schritten angegeben werden, fehlt noch die passende Skalierung:

2985	: O3EInt16(2, "ExternalHeaterTemperatureSetpoint", scale=10, signed=True), # 250-xH Unit °C

Passt auch zu meinen alten Aufzeichnungen aus der letzten Heizperiode...

Gruß Bu-Na

**Reply by cvbasta75** [reply]
OHNE scale=10 -> 70° MIT scale=10 -> 7.0° das kann nicht passen...

**Reply by bu-na** [reply]
Hmm, dann muss ich mal auf die nächste Anforderung warten und weiter testen, bei mit steht der Wert zur Zeit auf 0.0...

**Reply by MyHomeMyData** [reply]
Bei O3EInt16()  ist `scale=10` per Default gesetzt. Keine Angabe von `scale` und `scale=10` führen also zum gleichen Ergebnis. Siehe in `Open3Ecodecs.py`:
```
class O3EInt16(O3EInt):
    def __init__(self, string_len: int, idStr: str, scale: float = 10.0, offset: int = 0, signed=False):
        O3EInt.__init__(self, string_len, idStr, byte_width=2, scale=scale, offset=offset, signed=signed)
```

**Reply by cvbasta75** [reply]
Alles gut! Hatte aus versehen 100 drinnen stehen!
